ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/permissions/PeerPermissions.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.permissions;
+
+import tech.pegasys.pantheon.ethereum.p2p.peers.Peer;
+import tech.pegasys.pantheon.util.Subscribers;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class PeerPermissions {
+  private final Subscribers<PermissionsUpdateCallback> updateSubscribers = new Subscribers<>();
+
+  public static final PeerPermissions NOOP = new NoopPeerPermissions();
+
+  public static PeerPermissions noop() {
+    return NOOP;
+  }
+
+  public static PeerPermissions combine(final PeerPermissions... permissions) {
+    return combine(Arrays.asList(permissions));
+  }
+
+  public static PeerPermissions combine(final List<PeerPermissions> permissions) {
+    return CombinedPeerPermissions.create(permissions);
+  }
+
+  /**
+   * @param peer The {@link Peer} object representing the remote node
+   * @return True if we are allowed to communicate with this peer.
+   */
+  public abstract boolean isPermitted(final Peer peer);
+
+  public void subscribeUpdate(final PermissionsUpdateCallback callback) {
+    updateSubscribers.subscribe(callback);
+  }
+
+  protected void dispatchUpdate(
+      final boolean permissionsRestricted, final Optional<List<Peer>> affectedPeers) {
+    updateSubscribers.forEach(s -> s.onUpdate(permissionsRestricted, affectedPeers));
+  }
+
+  private static class NoopPeerPermissions extends PeerPermissions {
+    @Override
+    public boolean isPermitted(final Peer peer) {
+      return true;
+    }
+  }
+
+  private static class CombinedPeerPermissions extends PeerPermissions {
+    private final List<PeerPermissions> permissions;
```

Comment:
Consider ImmutableList.  Since we are allowing a list to shortcut other CombinedPeerPermissions by collapsing it into this list we should make it explicit that the object we are shortcutting won't mutate.  There is an ImmutableList collector as part of Guava.